### PR TITLE
[Doc Improvement][Update/add SSO for message extension]

### PR DESCRIPTION
**Pull request by MuyangAmigo**

### CODE REVIEW

**Comment by acrolinxatmsft1** [comment]
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/toolkit/add-single-sign-on.md](https://github.com/MicrosoftDocs/msteams-docs/blob/13d1e354e503c411d01a098d97d3a7422d122120/msteams-platform%2Ftoolkit%2Fadd-single-sign-on.md) | 89 | 13 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/be5b3e8d-8c62-4e95-936d-30df194da4e1) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 1]
```diff
@@ -18,10 +18,11 @@ For applications that interact with the user in a chat, a Team, or a channel, SS
...
 
-* Tab
-* Bot
-* Notification bot: restify server
-* Command bot
+- Tab
+- Bot
+- Notification bot: restify server
+- Command bot
+- Message extension
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:25]:
```suggestion
* Tab
* Bot
* Notification bot: restify server
* Command bot
* Message extension
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 2]
```diff
@@ -31,13 +32,13 @@ The following steps help you to add SSO using Teams Toolkit in Visual Studio Cod
 2. Select Teams Toolkit :::image type="content" source="~/assets/images/teams-toolkit-v2/teams-toolkit-sidebar-icon.png" alt-text="sso add sidebar"::: from left navigation bar.
 3. Select **Add features** under **DEVELOPMENT**.
 
-    :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="sso add features":::
+   :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="sso add features":::
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:35]:
```suggestion
:::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="Screenshot shows the Add features option under the Developemnt option in the Teams Toolkit.":::
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 3]
```diff
@@ -31,13 +32,13 @@ The following steps help you to add SSO using Teams Toolkit in Visual Studio Cod
...
 
-    * You can also open command palette and select **Teams: Add features**
+   - You can also open command palette and select **Teams: Add features**
 
 4. Select **Single Sign-On**.
 
-    :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-select features.png" alt-text="sso select":::
+   :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-select features.png" alt-text="sso select":::
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:41]:
```suggestion
:::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="Screenshot shows the Single Sign-on feature highlighted in red in the Teams Toolkit.":::
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 4]
```diff
@@ -68,28 +69,29 @@ The following steps help you to enable SSO in your application:
...
 
 ---
+
 <br>
 <br><details>
 <summary><b>Tab project
 </b></summary>
 
-1. Copy `auth-start.html` and `auth-end.htm`** in `auth/public` folder to `tabs/public/`. Teams Toolkit registers these two endpoints in Azure AD for Azure AD's redirect flow.
+1. Copy `auth-start.html` and `auth-end.htm`\*\* in `auth/public` folder to `tabs/public/`. Teams Toolkit registers these two endpoints in Azure AD for Azure AD's redirect flow.
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:78]: As per guidelines to write a folder path: In command syntax, use path to represent only the folder portion of the full path: copy [drive:][path]filename.

Request you to check and update the file path names.

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 5]
```diff
@@ -68,28 +69,29 @@ The following steps help you to enable SSO in your application:
...
 
 2. Copy `sso` folder under `auth/tab` to `tabs/src/sso/`.
 
-    * `InitTeamsFx`: The file implements a function that initializes TeamsFx SDK and opens `GetUserProfile` component after SDK is initialized
+   - `InitTeamsFx`: The file implements a function that initializes TeamsFx SDK and opens `GetUserProfile` component after SDK is initialized
 
-    * `GetUserProfile`: The file implements a function that calls Microsoft Graph API to get user info
+   - `GetUserProfile`: The file implements a function that calls Microsoft Graph API to get user info
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:84]:
```suggestion
*  `InitTeamsFx`: The file implements a function that initializes TeamsFx SDK and opens `GetUserProfile` component after SDK is initialized.

*  `GetUserProfile`: The file implements a function that calls Microsoft Graph API to get user info.
```

Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:84]:
```suggestion
   * `InitTeamsFx`: The file implements a function that initializes TeamsFx SDK and opens `GetUserProfile` component after SDK is initialized

   * `GetUserProfile`: The file implements a function that calls Microsoft Graph API to get user info
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 6]
```diff
@@ -68,28 +69,29 @@ The following steps help you to enable SSO in your application:
...
 
 3. Execute `npm install @microsoft/teamsfx-react` under `tabs/`.
 
 4. Add the following lines to `tabs/src/components/sample/Welcome.tsx` to import `InitTeamsFx`:
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:88]: As per guidelines to write a folder path: In command syntax, use path to represent only the folder portion of the full path: copy [drive:][path]filename.

Please check and update the file path.

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 7]
```diff
@@ -125,236 +127,295 @@ SSO command handler `ProfileSsoCommandHandler` uses an Azure AD token to call Mi
...
-                        }
-                    ]
-                };
-                return photoMessage;
-            } else {
-                return "Could not retrieve your photo from Microsoft Graph. Please make sure you have uploaded your photo.";
-            }
-        }
-    }
+1. Add `PhotoSsoCommandHandler` instance to `ssoCommands` array in `bot/src/internal/initialize.ts`:
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:321]: As per guidelines to write a folder path: In command syntax, use path to represent only the folder portion of the full path: copy [drive:][path]filename.

Request you to check and update the file path.

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 8]
```diff
@@ -125,236 +127,295 @@ SSO command handler `ProfileSsoCommandHandler` uses an Azure AD token to call Mi
...
-            ssoCommands: [new ProfileSsoCommandHandler(), new PhotoSsoCommandHandler()]
-        },
-    });
-    ...
+You can update the query logic in the `handleMessageExtensionQueryWithToken` with token which is obtained by using the logged-in Teams user token.
 
+To make this work in your application:
+1. Move the `auth/bot/public` folder to `bot`. This folder contains HTML pages that the bot application hosts. When single sign-on flows are initiated with AAD, AAD will redirect the user to these pages.
+
+1. Modify your `bot/index` to add the appropriate `restify` routes to these pages.
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:376]: As per guidelines to write a folder path: In command syntax, use path to represent only the folder portion of the full path: copy [drive:][path]filename.

Request you to check and update the file path across the doc.

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 9]
```diff
@@ -125,236 +127,295 @@ SSO command handler `ProfileSsoCommandHandler` uses an Azure AD token to call Mi
...
-    ```JSON
+1. Open `bot/package.json`, ensure that `@microsoft/teamsfx` version >= 1.2.0
 
-    {
-        "title": "photo",
-        "description": "Show user photo using Single Sign On feature"
-    }
+1. Install `isomorphic-fetch` npm packages in your bot project.
 
+1. (For ts only) Install `copyfiles` npm packages in your bot project, add or update the `build` script in `bot/package.json` as following
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:395]:
```suggestion
1. (For ts only) Install `copyfiles` npm packages in your bot project, add or update the `build` script in `bot/package.json` as follows:
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 10]
```diff
@@ -31,13 +32,13 @@ The following steps help you to add SSO using Teams Toolkit in Visual Studio Cod
 2. Select Teams Toolkit :::image type="content" source="~/assets/images/teams-toolkit-v2/teams-toolkit-sidebar-icon.png" alt-text="sso add sidebar"::: from left navigation bar.
 3. Select **Add features** under **DEVELOPMENT**.
 
-    :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="sso add features":::
+   :::image type="content" source="../assets/images/teams-toolkit-v2/add-sso/sso-add features.png" alt-text="sso add features":::
 
-    * You can also open command palette and select **Teams: Add features**
+   - You can also open command palette and select **Teams: Add features**
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:37]:
```suggestion
* You can also open command palette and select **Teams: Add features**.
```

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 11]
```diff
@@ -125,236 +127,295 @@ SSO command handler `ProfileSsoCommandHandler` uses an Azure AD token to call Mi
...
+     "title": "profile",
+     "description": "Show user profile using Single Sign On feature"
+   }
+   ```
 
 #### Add a new SSO command to the bot (Optional)
 
 After successfully adding SSO in your project, you can add a new SSO command.
 
 1. Create a new file such as `photoSsoCommandHandler.ts` or `photoSsoCommandHandler.js` in `bot/src/` and add your own SSO command handler to call Graph API:
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:209]: Is the code after this sentence related to this line. If yes, add a line about the code. Something in the lines of the following is an example of.... sample code.

File: msteams-platform/toolkit/add-single-sign-on.md [hunk 12]
```diff
@@ -125,236 +127,295 @@ SSO command handler `ProfileSsoCommandHandler` uses an Azure AD token to call Mi
...
+            directory: path.join(__dirname, "public"),
+        })
+    );
     ```
 
-1. Register your command in the Teams app manifest. Open `templates/appPackage/manifest.template.json`, and add following lines under `commands` in `commandLists` of your bot:
+1. Override `handleTeamsMessagingExtensionQuery` interface under `bot/teamsBot`. You can follow the sample code in the `handleMessageExtensionQueryWithToken` to do your own query logic.
 
-    ```JSON
+1. Open `bot/package.json`, ensure that `@microsoft/teamsfx` version >= 1.2.0
```
Comment (v-ypalikila) [inline msteams-platform/toolkit/add-single-sign-on.md:391]:
```suggestion
1. Open `bot/package.json`, ensure that `@microsoft/teamsfx` version >= 1.2.0.
```

**Comment by MuyangAmigo** [comment]
@v-ypalikila Thanks for the comments, are you going to help update the PR as well?

**Comment by acrolinxatmsft1** [comment]
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/toolkit/add-single-sign-on.md](https://github.com/MicrosoftDocs/msteams-docs/blob/b9e2c3e67b4824369acde6b3306abef5b4214a64/msteams-platform%2Ftoolkit%2Fadd-single-sign-on.md) | 92 | 10 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/8901e5a3-0efb-4fe6-bd64-9d342f3711fa) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

**Comment by acrolinxatmsft1** [comment]
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/toolkit/add-single-sign-on.md](https://github.com/MicrosoftDocs/msteams-docs/blob/9076b50c321b2cab908eae2cf84610a27bba01d4/msteams-platform%2Ftoolkit%2Fadd-single-sign-on.md) | 92 | 10 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/d9c4e4d6-7834-4135-a001-4fc549eb70f5) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 1 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md]: updated in the PR.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 2 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md]: updated in the PR.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 7 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md:321]: As discussed with the PM, the PM explained that we can't add a dive path as it depends on where the user installs the package. He also confirmed that the folder structure remains the same.

Change not required.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 8 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md]: As discussed with the PM, the PM explained that we can't add a dive path as it depends on where the user installs the package. He also confirmed that the folder structure remains the same.

Change not required.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 10 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md]: updated in the PR.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 12 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md:392]: Updated in the PR.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 9 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md]: updated in PR.

File: msteams-platform/toolkit/add-single-sign-on.md (hunk 6 above)
Comment (v-ypalikila) [reply msteams-platform/toolkit/add-single-sign-on.md:88]: As discussed with the PM, the PM explained that we can't add a dive path as it depends on where the user installs the package. He also confirmed that the folder structure remains the same.

Change not required.

**Pull request merged**